File: camera/camera2/src/main/java/androidx/camera/camera2/Camera2CameraInfo.java
/*
 * Copyright (C) 2019 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package androidx.camera.camera2;

import android.annotation.SuppressLint;
import android.hardware.camera2.CameraAccessException;
import android.hardware.camera2.CameraCharacteristics;
import android.hardware.camera2.CameraManager;
import android.view.Surface;

import androidx.annotation.Nullable;
import androidx.camera.core.CameraInfo;
import androidx.camera.core.CameraInfoUnavailableException;
import androidx.camera.core.CameraOrientationUtil;
import androidx.camera.core.CameraX.LensFacing;
import androidx.camera.core.ImageOutputConfiguration.RotationValue;
import androidx.core.util.Preconditions;

/** Implementation of the {@link CameraInfo} interface that exposes parameters through camera2. */
final class Camera2CameraInfo implements CameraInfo {

    private final CameraCharacteristics mCameraCharacteristics;

    Camera2CameraInfo(CameraManager cameraManager, String cameraId)
            throws CameraInfoUnavailableException {
        try {
            mCameraCharacteristics = cameraManager.getCameraCharacteristics(cameraId);
        } catch (CameraAccessException e) {
            throw new CameraInfoUnavailableException(
                    "Unable to retrieve info for camera " + cameraId, e);
        }

        checkCharacteristicAvailable(
                CameraCharacteristics.SENSOR_ORIENTATION, "Sensor orientation");
        checkCharacteristicAvailable(CameraCharacteristics.LENS_FACING, "Lens facing direction");
    }

    @SuppressLint("RestrictedApi") // TODO(b/124323692): Remove after aosp/900913 is merged
    @Nullable
    @Override
    public LensFacing getLensFacing() {
        Integer lensFacing = mCameraCharacteristics.get(CameraCharacteristics.LENS_FACING);
        Preconditions.checkNotNull(lensFacing);
        switch (lensFacing) {
            case CameraCharacteristics.LENS_FACING_FRONT:
                return LensFacing.FRONT;
            case CameraCharacteristics.LENS_FACING_BACK:
                return LensFacing.BACK;
            default:
                return null;
        }
    }

    @SuppressLint("RestrictedApi") // TODO(b/124323692): Remove after aosp/900913 is merged
    @Override
    public int getSensorRotationDegrees(@RotationValue int relativeRotation) {
        Integer sensorOrientation =
                mCameraCharacteristics.get(CameraCharacteristics.SENSOR_ORIENTATION);
        Preconditions.checkNotNull(sensorOrientation);
        int relativeRotationDegrees =
                CameraOrientationUtil.surfaceRotationToDegrees(relativeRotation);
        // Currently this assumes that a back-facing camera is always opposite to the screen.
        // This may not be the case for all devices, so in the future we may need to handle that
        // scenario.
        boolean isOppositeFacingScreen = LensFacing.BACK.equals(getLensFacing());
        return CameraOrientationUtil.getRelativeImageRotation(
                relativeRotationDegrees,
                sensorOrientation,
                isOppositeFacingScreen);
    }

    private void checkCharacteristicAvailable(CameraCharacteristics.Key<?> key, String readableName)
            throws CameraInfoUnavailableException {
        if (mCameraCharacteristics.get(key) == null) {
            throw new CameraInfoUnavailableException(
                    "Camera characteristics map is missing value for characteristic: "
                            + readableName);
        }
    }

    @Override
    public int getSensorRotationDegrees() {
        return getSensorRotationDegrees(Surface.ROTATION_0);
    }
}

File: camera/camera-camera2-pipe-integration/src/main/java/androidx/camera/camera2/pipe/integration/interop/Camera2CameraInfo.kt
/*
 * Copyright 2022 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package androidx.camera.camera2.pipe.integration.interop

import android.hardware.camera2.CameraCharacteristics
import androidx.annotation.RequiresApi
import androidx.camera.core.CameraInfo

/**
 * An interface for retrieving Camera2-related camera information.
 */
@ExperimentalCamera2Interop
@RequiresApi(21) // TODO(b/200306659): Remove and replace with annotation on package-info.java
class Camera2CameraInfo private constructor() {

    /**
     * Gets a camera characteristic value.
     *
     * The characteristic value is the same as the value in the [CameraCharacteristics]
     * that would be obtained from
     * [android.hardware.camera2.CameraManager.getCameraCharacteristics].
     *
     * @param <T> The type of the characteristic value.
     * @param key The [CameraCharacteristics.Key] of the characteristic.
     * @return the value of the characteristic.
    </T> */
    fun <T> getCameraCharacteristic(
        @Suppress("UNUSED_PARAMETER") key: CameraCharacteristics.Key<T>
    ): T? {
        TODO()
    }

    /**
     * Gets the string camera ID.
     *
     *
     * The camera ID is the same as the camera ID that would be obtained from
     * [android.hardware.camera2.CameraManager.getCameraIdList]. The ID that is retrieved
     * is not static and can change depending on the current internal configuration of the
     * [androidx.camera.core.Camera] from which the CameraInfo was retrieved.
     *
     * The Camera is a logical camera which can be backed by multiple
     * [android.hardware.camera2.CameraDevice]. However, only one CameraDevice is active at
     * one time. When the CameraDevice changes then the camera id will change.
     *
     * @return the camera ID.
     * @throws IllegalStateException if the camera info does not contain the camera 2 camera ID
     * (e.g., if CameraX was not initialized with a
     * [androidx.camera.camera2.Camera2Config]).
     */

    fun getCameraId(): String {
        TODO()
    }

    /**
     * @hide
     */
    companion object {

        /**
         * Gets the [Camera2CameraInfo] from a [CameraInfo].
         *
         * @param cameraInfo The [CameraInfo] to get from.
         * @return The camera information with Camera2 implementation.
         * @throws IllegalArgumentException if the camera info does not contain the camera2 information
         * (e.g., if CameraX was not initialized with a
         * [androidx.camera.camera2.Camera2Config]).
         */
        @JvmStatic
        fun from(@Suppress("UNUSED_PARAMETER") cameraInfo: CameraInfo): Camera2CameraInfo {
            TODO()
        }
    }
}